src/main/java/org/myrobotlab/service/RasPi.java

```diff
@@ -376,6 +541,110 @@ public PinData publishPin(PinData pinData) {
     return pinData;
   }
 
+  public Map<Integer, Set<String>> publishScan() {
+    return validI2CAddresses;
+  }
+
+  public void read() {
+    log.debug("read task invoked");
+    List<PinData> pinArray = new ArrayList<>();
+    // load pin array
+    for (String pin : pinIndex.keySet()) {
+      PinDefinition pindef = pinIndex.get(pin);
+      if (pindef.isEnabled()) {
+        log.debug("pin {} enabled {}", pin, pindef.isEnabled());
+        int value = read(pin);
+        pindef.setValue(value);
+        PinData pd = new PinData(pin, value);
+        log.debug("pin data {}", pd);
+        pinArray.add(pd);
+      }
+    }
+
+    if (pinArray.size() > 0) {
+      PinData[] array = pinArray.toArray(new PinData[0]);
+      invoke("publishPinArray", new Object[] { array });
+    }
+  }
+
+  @Override
+  public int read(String pin) {
+
+    if (!pinIndex.containsKey(pin)) {
+      error("Pin %s not found", pin);
+      return -1;
+    }
+    PinDefinition pindef = pinIndex.get(pin);
+    GpioPinDigitalMultipurpose gpioPin = getGPIO(pin);
+    if (!gpioPin.isMode(PinMode.DIGITAL_INPUT)) {
+      pinMode(pin, "INPUT");
+    }
+    if (gpioPin.isLow()) {
+      pindef.setValue(0);
+      return 0;
+    } else {
+      pindef.setValue(1);
+      return 1;
+    }
+  }
+
+  @Override
+  public void reset() {
+    // TODO Auto-generated method stub
+    // reset pins/i2c devices/gpio pins
+  }
+
+  public void scan() {
+    scan(null);
+  }
+
+  public void scan(Integer busNumber) {

```

Comment:
Probably needs javadoc. And is wrapper type needed or can primitive be used?

Reply:
I debated if the parameter should be taken out completely.  @Cyber-One informed me that raspi only exposes bus 1 for i2c .. so there is no choice